### ISSUE TITLE
Debugging a hart not robust across non-debugger controlled reset

### ISSUE THREAD

**Issue by mwachs5** [riscv/riscv-openocd#211]
Consider the case where openOCD is debugging a hart and one of its actions cause the hart to be reset (e.g writing a memory location which leads to soft reset), or someone presses an external reset button. OpenOCD doesn't re-halt the hart or otherwise gracefully recover (it works fine if OpenOCD knowingly caused the reset with `ndreset`).

Most of the problems could be avoided if OpenOCD left `haltreq` set until it actually wanted to resume. This would cause a hart which was being debugged and got reset to halt again even if it was reset out of the debugger's control.

**Comment by mwachs5** [comment]
#226 doesn't help this. It just tells the user, but it doesn't actually keep the hart halted (I think, please correct me if I'm wrong).

**Comment by timsifive** [comment]
You're right. It gets us closer, but doesn't implement what this issue asks for. I'll get to that.

**Comment by mwachs5** [comment]
#226 doesn't resolve this issue as noted in your comment above (I thought)

**Comment by timsifive** [comment]
Yes. I think github found a magic string in one of the #226 commits and automatically closed this issue.

**Comment by elfmimi** [comment]
I believe dmstatus should be monitored, at least fetched again, after writing halt request to dmcontrol.
https://github.com/riscv/riscv-openocd/blob/cbb15587dc782ac8ade7ae252e7b760cfba4a178/src/target/riscv/riscv-013.c#L3852-L3856
